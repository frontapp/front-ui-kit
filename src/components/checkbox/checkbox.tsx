import React, {FC, useLayoutEffect, useRef} from 'react';
import styled, {css} from 'styled-components';

import {Icon} from '../../elements/icon/icon';
import {greys, palette} from '../../helpers/colorHelpers';
import {fonts, fontSizes, fontWeights} from '../../helpers/fontHelpers';

/*
 * Props.
 */

interface CheckboxProps {
  /** Whether the checkbox is checked. */
  isChecked: boolean;
  /** The label of the checkbox. */
  children?: React.ReactNode;
  /** Whether the checkbox is disabled. If disabled, clicking it won't check/uncheck it. */
  isDisabled?: boolean;
  /** Whether the checkbox is in an indeterminate state. */
  isIndeterminate?: boolean;
  /** The side to render the label on. */
  labelSide?: 'left' | 'right';
  /** Handler to check or uncheck the checkbox */
  onChange: (isChecked: boolean) => void;
}

/*
 * Style.
 */

interface StyledCheckboxInputProps {
  $isChecked: boolean;
  $isIndeterminate: boolean;
  $isDisabled: boolean;
}

interface StyledWrapperDivProps {
  $labelSide: 'left' | 'right';
}

const StyledWrapperDiv = styled.div<StyledWrapperDivProps>`
  display: flex;
  flex-flow: row;
  align-items: center;

  ${p =>
    p.$labelSide === 'left' &&
    css`
      flex-flow: row-reverse;
    `};
`;

const StyledCheckboxDiv = styled.div`
  position: relative;
  width: 16px;
  height: 16px;
`;

const StyledIconDiv = styled.div`
  position: absolute;
  pointer-events: none;
  top: 0px;
`;

const StyledInput = styled.input<StyledCheckboxInputProps>`
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  appearance: none;
  margin: 0px;

  ${p => addCheckboxStyles(p)};
`;

const StyledChildrenDiv = styled.div`
  font-family: ${fonts.system};
  font-size: ${fontSizes.medium};
  font-weight: ${fontWeights.normal};
  line-height: 17px;
  color: ${greys.shade80};
  margin-left: 8px;
  cursor: default;
  user-select: none;
  flex: 1;

  &:empty {
    margin-left: 0;
    display: none;
  }
`;

function addCheckboxStyles(props: StyledCheckboxInputProps) {
  if (props.$isDisabled && (props.$isChecked || props.$isIndeterminate))
    return css`
      background-color: ${greys.shade40};
      border: 1.5px solid ${greys.shade40};
    `;

  if (props.$isDisabled)
    return css`
      background-color: ${greys.shade20};
      border: 1.5px solid ${greys.shade40};
    `;

  if (props.$isChecked || props.$isIndeterminate)
    return css`
      background: ${palette.blue.shade40};
      border: none;
    `;

  return css`
    border: 1.5px solid ${greys.shade60};
  `;
}

/*
 * Component.
 */

export const Checkbox: FC<CheckboxProps> = ({
  isChecked,
  children,
  isDisabled = false,
  isIndeterminate = false,
  labelSide = 'right',
  onChange
}) => {
  const onInputChange = () => {
    if (isDisabled)
      return;
    onChange(!isChecked);
  };
  const inputRef = useRef<HTMLInputElement>(null);

  // Set the attributes of the input on every render.
  useLayoutEffect(() => {
    const input = inputRef.current;
    if (!input)
      return;

    if (isDisabled)
      input.disabled = true;
    else
      input.disabled = false;

    if (isChecked) {
      input.indeterminate = false;
      input.checked = true;
    } else if (isIndeterminate) {
      input.indeterminate = true;
      input.checked = false;
    }
  }, [isDisabled, isChecked, isIndeterminate]);

  return (
    <StyledWrapperDiv $labelSide={labelSide}>
      <StyledCheckboxDiv>
        <StyledInput
          ref={inputRef}
          type="checkbox"
          $isChecked={isChecked}
          $isDisabled={isDisabled}
          $isIndeterminate={isIndeterminate}
          onChange={onInputChange}
        />
        <StyledIconDiv>{getCheckboxIcon(isDisabled, isChecked, isIndeterminate)}</StyledIconDiv>
      </StyledCheckboxDiv>
      <StyledChildrenDiv onClick={onInputChange}>{children}</StyledChildrenDiv>
    </StyledWrapperDiv>
  );
};

/*
 * Helpers.
 */

function getCheckboxIcon(isDisabled: boolean, isChecked: boolean, isIndeterminate: boolean) {
  if (isChecked)
    return <Icon name="CheckmarkBox" color={isDisabled ? greys.shade60 : greys.white} />;
  if (isIndeterminate)
    return <Icon name="Minus" color={isDisabled ? greys.shade60 : greys.white} />;
  return null;
}
