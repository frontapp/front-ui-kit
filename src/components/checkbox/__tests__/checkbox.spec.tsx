import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import {Checkbox} from '../checkbox';

describe('<Checkbox /> Spec', () => {
  describe('Snapshots', () => {
    it('should render an empty and disabled checkbox', () => {
      const view = render(<Checkbox onChange={jest.fn()} isChecked={false} isDisabled />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render a checked and disabled checkbox', () => {
      const view = render(<Checkbox onChange={jest.fn()} isChecked isDisabled />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an indeterminate and disabled checkbox', () => {
      const view = render(<Checkbox onChange={jest.fn()} isChecked={false} isIndeterminate isDisabled />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an empty checkbox', () => {
      const view = render(<Checkbox onChange={jest.fn()} isChecked={false} />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render a checked checkbox', () => {
      const view = render(<Checkbox onChange={jest.fn()} isChecked />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an indeterminate checkbox', () => {
      const view = render(<Checkbox onChange={jest.fn()} isIndeterminate isChecked={false} />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render a checkbox with a label', () => {
      const view = render(
        <Checkbox onChange={jest.fn()} isChecked={false}>
          Label
        </Checkbox>
      );
      expect(view.container).toMatchSnapshot();
    });
  });

  it('should issue an onChange event with an argument "true" when the checkbox is unchecked', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const view = render(
      <Checkbox onChange={onChangeMock} isChecked={false}>
        Label
      </Checkbox>
    );
    await user.click(view.getByRole('checkbox'));
    expect(onChangeMock).toBeCalled();
    expect(onChangeMock.mock.lastCall[0]).toBe(true);
  });

  it('should issue an onChange event with an argument "false" when the checkbox is checked', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const view = render(
      <Checkbox onChange={onChangeMock} isChecked>
        Label
      </Checkbox>
    );
    await user.click(view.getByRole('checkbox'));
    expect(onChangeMock).toBeCalled();
    expect(onChangeMock.mock.lastCall[0]).toBe(false);
  });

  it('should issue an onChange event with an argument "true" when the checkbox is indeterminate', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const view = render(
      <Checkbox onChange={onChangeMock} isChecked={false} isIndeterminate>
        Label
      </Checkbox>
    );
    await user.click(view.getByRole('checkbox'));
    expect(onChangeMock).toBeCalled();
    expect(onChangeMock.mock.lastCall[0]).toBe(true);
  });

  it('should not issue an onChange event when the checkbox is disabled', async () => {
    const user = userEvent.setup();
    const onChangeMock = jest.fn();
    const view = render(
      <Checkbox onChange={onChangeMock} isChecked isDisabled>
        Label
      </Checkbox>
    );
    await user.click(view.getByRole('checkbox'));
    expect(onChangeMock).not.toBeCalled();
  });
});
