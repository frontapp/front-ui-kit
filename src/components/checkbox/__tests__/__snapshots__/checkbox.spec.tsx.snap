// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> Spec Snapshots should render a checkbox with a label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  width: 16px;
  height: 16px;
}

.c3 {
  position: absolute;
  pointer-events: none;
  top: 0px;
}

.c2 {
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0px;
  border: 1.5px solid #9b9c9e;
}

.c4 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  color: #3e3e40;
  margin-left: 8px;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4:empty {
  margin-left: 0;
  display: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        type="checkbox"
      />
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
    >
      Label
    </div>
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render a checked and disabled checkbox 1`] = `
.c4 {
  width: 16px;
  height: 16px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  width: 16px;
  height: 16px;
}

.c3 {
  position: absolute;
  pointer-events: none;
  top: 0px;
}

.c2 {
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0px;
  background-color: #d4d6d9;
  border: 1.5px solid #d4d6d9;
}

.c5 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  color: #3e3e40;
  margin-left: 8px;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5:empty {
  margin-left: 0;
  display: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        disabled=""
        type="checkbox"
      />
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            color="#9b9c9e"
            height="16"
            name="CheckmarkBox"
            preserveaspectratio="none"
            width="16"
          />
        </div>
      </div>
    </div>
    <div
      class="c5"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render a checked checkbox 1`] = `
.c4 {
  width: 16px;
  height: 16px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  width: 16px;
  height: 16px;
}

.c3 {
  position: absolute;
  pointer-events: none;
  top: 0px;
}

.c2 {
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0px;
  background: #367fee;
  border: none;
}

.c5 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  color: #3e3e40;
  margin-left: 8px;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5:empty {
  margin-left: 0;
  display: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        type="checkbox"
      />
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            color="#ffffff"
            height="16"
            name="CheckmarkBox"
            preserveaspectratio="none"
            width="16"
          />
        </div>
      </div>
    </div>
    <div
      class="c5"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an empty and disabled checkbox 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  width: 16px;
  height: 16px;
}

.c3 {
  position: absolute;
  pointer-events: none;
  top: 0px;
}

.c2 {
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0px;
  background-color: #f3f4f6;
  border: 1.5px solid #d4d6d9;
}

.c4 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  color: #3e3e40;
  margin-left: 8px;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4:empty {
  margin-left: 0;
  display: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        disabled=""
        type="checkbox"
      />
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an empty checkbox 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  width: 16px;
  height: 16px;
}

.c3 {
  position: absolute;
  pointer-events: none;
  top: 0px;
}

.c2 {
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0px;
  border: 1.5px solid #9b9c9e;
}

.c4 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  color: #3e3e40;
  margin-left: 8px;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4:empty {
  margin-left: 0;
  display: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        type="checkbox"
      />
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an indeterminate and disabled checkbox 1`] = `
.c4 {
  width: 16px;
  height: 16px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  width: 16px;
  height: 16px;
}

.c3 {
  position: absolute;
  pointer-events: none;
  top: 0px;
}

.c2 {
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0px;
  background-color: #d4d6d9;
  border: 1.5px solid #d4d6d9;
}

.c5 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  color: #3e3e40;
  margin-left: 8px;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5:empty {
  margin-left: 0;
  display: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        disabled=""
        type="checkbox"
      />
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            color="#9b9c9e"
            height="16"
            name="Minus"
            preserveaspectratio="none"
            width="16"
          />
        </div>
      </div>
    </div>
    <div
      class="c5"
    />
  </div>
</div>
`;

exports[`<Checkbox /> Spec Snapshots should render an indeterminate checkbox 1`] = `
.c4 {
  width: 16px;
  height: 16px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  width: 16px;
  height: 16px;
}

.c3 {
  position: absolute;
  pointer-events: none;
  top: 0px;
}

.c2 {
  width: inherit;
  height: inherit;
  border-radius: 3px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0px;
  background: #367fee;
  border: none;
}

.c5 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 14px;
  font-weight: 400;
  line-height: 17px;
  color: #3e3e40;
  margin-left: 8px;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5:empty {
  margin-left: 0;
  display: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        type="checkbox"
      />
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            color="#ffffff"
            height="16"
            name="Minus"
            preserveaspectratio="none"
            width="16"
          />
        </div>
      </div>
    </div>
    <div
      class="c5"
    />
  </div>
</div>
`;
